### ISSUE TITLE
[Homepage email signup] Create React widget for homepage email signup form and add new error handling states

### ISSUE THREAD

**Issue by randimays** [department-of-veterans-affairs/va.gov-cms#19516]
## Description
On VA.gov at the bottom of the homepage, there is a section titled "Sign up to get the latest VA updates" with an input for an email address.

There is currently no error handling in place for this form. If the form fails to reach the GovDelivery service, the user will be directed to a browser error page (404). There is also no input validation on the form, allowing a user to submit the form with a blank email address or with any combination of characters.

We need to address these gaps in error handling, but we want to avoid adding a lot of Javascript (client-side code) to content-build. Since the form currently lives exclusively in content-build, we need to create a React widget in vets-website for the form and implement it on the homepage in content-build.

Example of an existing vets-website widget:
- [Language toggle - vets-website](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/static-pages/i18Select)
- [Widget implementation (empty div) - content-build](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/site/layouts/support_resources_detail_page.drupal.liquid#L13)

### Email signup form, present and future

<details><summary>Existing design (production)</summary>

#### Mobile
<img src="https://github.com/user-attachments/assets/4cf81216-5f2f-46c4-b772-829529190459" width="350" />

#### Desktop
<img src="https://github.com/user-attachments/assets/4886aac0-1580-4afd-a206-697c4a8bfa1d" width="700" />

</details>

<details><summary>New design</summary>

In #19314, DaveP created the error state for when the GovDelivery service rejects our request or the server response fails for any reason.

Figma: https://www.figma.com/design/60PPoOSpKiXwmNHT01C09J/Home-Page-Updates?node-id=38-3849

#### Input validation states - desktop and mobile
<img src="https://github.com/user-attachments/assets/3fd8217b-f6a0-4bff-bcab-d8da65407d80" width="650" />

#### Server issues - mobile

<img src="https://github.com/user-attachments/assets/e3be1beb-5440-4ae7-8b39-c3119881f00d" width="350" />

#### Server issues - desktop

<img src="https://github.com/user-attachments/assets/703137d6-8333-45a3-96d6-7952d128b67e" width="800" />

</details>

### Summary of the tasks for this ticket
- (content-build) Add a feature toggle to control display of the static HTML form (CB) or a new empty container pointing to a new React widget in vets-website
- (vets-website) Create a React widget for the form to include:
  - Input validation built in from the web component(s)
  - Input validation messaging aligning with the Figma design in `New design` above
  - ~Server error handling (pink banner) per the Figma design~ We do not have access to or control over the GovDelivery server response, so we don't have a use case for this banner now
- Add unit ~and/or e2e~ tests for the new React widget as appropriate / capable. Note: React widgets are not able to be tested because the page is split between content-build and vets-website and this creates issues testing in any environment.
- Update [monitoring documentation](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/home-page/engineering/monitoring.md#homepage-email-signup-govdelivery-call-is-failing)
- Update [Datadog homepage synthetic](https://vagov.ddog-gov.com/synthetics/details/gsg-zui-hg8) - this has a check for the email input

### User story

**AS A** Veteran or other user when on the VA.gov home page signing up for emailed VA updates
**I WANT** to know if I make a mistake when entering or submitting my email address
**SO THAT** I can correct the mistakes and there isn't a 'silent' failure of my signup, thus resulting in my not receiving email updates.

### Engineering notes / background / launch plan
It may be helpful to read the technical notes put together as part of the Zero Silent Failures SPIKE for this component: https://github.com/department-of-veterans-affairs/va.gov-cms/issues/19242#issuecomment-2372312641

This is a cutover from content-build content to vets-website ticket. 

- [x] ~Create a vets-api Flipper~ We discussed not needing this feature flipper because the vets-website widget will be looking for a specific container coming from content-build. If it is not there, the widget will not do anything (i.e. there is nothing to flip)
- [x] Create CMS feature toggle, enabled by default
- [x] In content-build: put existing content behind the feature toggle, where enabled = content build content shows, disabled = content-build content does not show.
- [x] Content-build: merge change to keep content-build content visible if CMS toggle is on, wait for prod deploy
- [x] FE: Merge vets-website, wait for prod deploy
- [x] In a Tugboat: test, verify, and document the order of steps to: 
   - [x] Run QA test plan and verify state
   - [x] Turn off the content build visibility
   - [x] Turn on the vets-website build visibility
   - [x] Verify new form design

### Analytics considerations
// ~TODO~ DONE (see below) @FranECross: We currently have a custom event on the submit button for this form. It logs two events:
```
{
  event: 'cta-button-click',
  'button-type': 'primary',
  'button-click-label': 'Sign up'
}

{
  event: 'homepage-email-sign-up',
  action: 'Homepage email signup'
}
```

If we choose to use the Design System analytics for this button, we'll get this one event:
```
{
  type: 'primary',
  label: 'Sign up'
}
```

@FranECross: The custom event clearly has more detail, but do you want to change anything about it as part of this ticket? OR keep the custom event as-is? **_Fran: Let's keep the custom event, unchanged. Thanks!_**

11/1: Fran & Randi discussed whittling this down to a single event instead:
```
event: 'homepage-email-sign-up'
button-click-label: 'Sign up'
```

## Testing & QA
* @FranECross to help with regression test plan
* At minimum, don't regress ability to signup via GovDelivery form

### Roles / assignments
_After functional testing, code review, accessibility review, and design review can happen in parallel._
- [ ] **Functional testing:** Chris Kim, Fran Cross (if needed)
- [x] **Code review:** Chris Kim
- [ ] **Design review:** @davidmpickett 
- [ ] **Accessibility review:** @laflannery 
- [ ] **Product acceptance:** Fran Cross

## Acceptance criteria
_ACs should capture expected behavior, to inform test plans. Consider devices, documentation updates including KBs, change management, and [content model](https://prod.cms.va.gov/admin/structure/cm_document) when applicable._
- [x] Put existing GovDelivery / email signup HTML form on a feature toggle (it should show when the toggle is ON)
- [x] Add an empty container pointing to a new React widget in vets-website (it should show when the toggle is OFF)
- [x] Create a React widget in vets-website for the form to include:
  - [x] Input validation built in from the web component(s)
  - [x] Input validation messaging aligning with the Figma design in `New design` above
  - [x] ~Server error handling (pink banner) per the Figma design~ We do not have access to or control over the GovDelivery server response, so we don't have a use case for this banner now
- [x] Add unit and/or e2e tests for the new React widget as appropriate / capable
- [x] Update [Datadog homepage synthetic](https://vagov.ddog-gov.com/synthetics/details/gsg-zui-hg8) - this has a check for the email input
- [x] Email form scales appropriately for different browser font sizes
- [x] Create followup ticket for removing feature toggle. #19888
- [x] Accessibility review
- [x] Design review
- [x] Impact analysis is filled out and QA test plan is updated accordingly
- [x] Does this product have an existing regression test plan?
	- [x] Yes, link: https://docs.google.com/spreadsheets/d/1pxzl5N3SDQIbckJnyigU1RbK_qIGMKKCN0ElksuffxM/edit?gid=0#gid=0
	- [ ] No, regression test plan will be created by Fran Cross as part of the work / points in this ticket

**Comment by FranECross** [comment]
@randimays I've answered the analytics question, added names to the roles, and also ticked the box for no current regression test, but indicated that I will create one for this issue. I'll get it done before the work begins. cc

**Comment by jilladams** [comment]
Note: @FranECross test plan needs to move to the team Test Plans folder

**Comment by FranECross** [comment]
> Note: [@FranECross](https://github.com/FranECross) test plan needs to move to the team Test Plans folder

Moved; thanks for the reminder!

**Comment by randimays** [comment]
@davidmpickett We also need to warn of a 130 maximum character limit on the email address - I just discovered this in my testing the GovDelivery form. Could you please provide some error text for this scenario?

**Comment by FranECross** [comment]
> [@davidmpickett](https://github.com/davidmpickett) We also need to warn of a 130 maximum character limit on the email address - I just discovered this in my testing the GovDelivery form. Could you please provide some error text for this scenario?

@randimays I've updated the regression plan with a line item for the 130 character limit, and will update with more details after Dave provides error messaging.

**Comment by davidmpickett** [comment]
So fields with character limits should use the [character count variant](https://design.va.gov/components/form/text-input#character-count) of the text input component. This displays a live countdown of characters remains and prevents the user from entering more characters than the limit.

**Comment by davidmpickett** [comment]
@randimays Let's just go with the more general message "Enter a valid email address without spaces using this format: email@domain.com"

**Comment by randimays** [comment]
Update: CMS feature toggle (`FEATURE_EMAIL_UPDATE_WIDGET`) is merged and is ON in both production and staging. I created [a Tugboat](https://cms-g1rglcuobmlvgewnhny9uo7f6vigntwi.demo.cms.va.gov/) and the feature toggle is OFF in that environment.

[PR for content-build](https://github.com/department-of-veterans-affairs/content-build/pull/2342) feature toggle & container for the React widget is tested & approved by our team and awaiting Platform review.

The code for the new React widget is complete and has been tested locally. I'm waiting for the content-build PR to be merged so I can test vets-website against content-build's `main` and validate the feature toggle code in production before merging the widget.

**Comment by randimays** [comment]
The CI for the content-build code finally passed today (I think it was just luck / Platform folks re-running while I was out). I verified the code working properly in staging and production. The feature toggle is ON and the current/existing form is still functioning as expected in both environments. Next step is for me to re-test the new widget code locally and work on getting that merged.

**Comment by randimays** [comment]
@davidmpickett @FranECross It turns out we won't be able to handle errors for the service call to GovDelivery. Since we're needing to prefill the email address, we have to structure the service call a specific way that sends the request to them without having access to the response. This means we won't be able to implement the `We've run into a problem` banner. GovDelivery will handle any service failures on their end / site. We can still do input validation, so all of that error handling / character counting has been added.

**Comment by FranECross** [comment]
Thanks for this information, @randimays ! I'd like to defer to Dave P for the UX question. Regarding the inability to handle errors for GovDelivery, I feel like that's okay, since it won't be a silent error to the user - they'll just see the errors on the GovDelivery site. @jilladams do you concur?

**Comment by randimays** [comment]
End-of-sprint update: The widget code (vets-website) is approved by our team and needs a Platform review. Once merged, I will flip the feature toggle off so we can test in staging -- we need a design and accessibility review there. Once approved, we can go to production, flip the feature toggle off there and validate that everything is working as expected. This will definitely carry over.

**Comment by randimays** [comment]
Update: React widget code was merged a couple hours ago. Prepping a Tugboat for testing for design and a11y. Once that is good to go, we can turn the feature toggle off in production and test there (early next week).

**Comment by randimays** [comment]
@FranECross Ticket for removing the feature toggle and related code has been created: https://github.com/department-of-veterans-affairs/va.gov-cms/issues/19888. It is a little stubby and could use some Product eyes 👀 

After this ticket is complete, we can pull in the above ticket when ready.

**Comment by laflannery** [comment]
@randimays I don't know where the PR for this is so I'm putting my comments here instead:
1. The gray hint text `(rgb(113, 118, 122)` against the blue background `#e1f3f8` does not pass contract requirements. It's only 4.02:1
![Image](https://github.com/user-attachments/assets/cf8fa103-cd3a-4adf-b154-2d768077e7de)
   - I'm also noticing that in the Figma we don't have the character counter - was this a later addition?
2. Focus management: when the error appears, the next tab stop should be the input.
    -  We do this on PACT and DUW and it's easier because the error doesn't appear on blur.
    - If you force focus to the error message, this should solve the issue

**Comment by randimays** [comment]
@davidmpickett we have a contrast issue with the character counter text against the blue background. See Laura's comment above. Do you have a preferred darker color to change this text to?

@laflannery The character counter was decided on in the comments of this ticket [here](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/19516#issuecomment-2450624098). I'll work on a fix for the focus issue.

**Comment by laflannery** [comment]
@randimays I looked at the [design system color palette](https://design.va.gov/foundation/color-palette#component-color-tokens) and there weren't a ton of gray options that worked but I did find one `vads-color-base-dark`. That's a ratio of 5.9:1. Are you able to change the counter to this?

**Comment by randimays** [comment]
@laflannery Yes, I'll do that!

**Comment by laflannery** [comment]
Color contrast looks good! Focus is also good but I want to run it with a SR just in case and I am currently in meetings so I will have to do this later this afternoon

**Comment by randimays** [comment]
Design approved the widget in Tugboat. New email widget is live in production. Email synthetic has been updated with the new selector in Datadog. Closing as complete.
